### PR TITLE
Service Fabric server certificate common names

### PR DESCRIPTION
**Pull request by steve-hawkins**
certificate common names can be used in place of thumbprints as recommended for production clusters

makes certificate rotation process faster / safer as the manifest does not require updating and a rolling upgrade

more info on the [Microsoft docs](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-create-cluster-using-cert-cn)

### CODE REVIEW

File: azurerm/resource_arm_service_fabric_cluster.go [hunk 1]
```diff
@@ -644,6 +687,68 @@ func flattenServiceFabricClusterCertificate(input *servicefabric.CertificateDesc
...
+	}
+
+	output := make(map[string]interface{})
+
+	if commonNames := in.CommonNames; commonNames != nil {
+		common_names := make([]map[string]interface{}, 0, len(*commonNames))
+		for _, i := range *commonNames {
+			commonName := make(map[string]interface{})
+
+			commonName["certificate_common_name"] = *i.CertificateCommonName
```
Comment (mbfrahry) [inline azurerm/resource_arm_service_fabric_cluster.go]: We should nil check this

**Review by mbfrahry — CHANGES_REQUESTED** [review]
Hey @steve-hawkins. These changes are looking good. I've got a few minor comments and one of the tests isn't passing.

```
--- FAIL: TestAccAzureRMServiceFabricCluster_readerAdminClientCertificateThumbprint (62.20s)
    testing.go:568: Step 0 error: Check failed: Check 5/12 error: azurerm_service_fabric_cluster.test: Attribute 'client_certificate_thumbprint.#' expected "2", got "0"
```

We should be able to get this merged in once these comments have been addressed

File: azurerm/resource_arm_service_fabric_cluster.go [hunk 2]
```diff
@@ -644,6 +687,68 @@ func flattenServiceFabricClusterCertificate(input *servicefabric.CertificateDesc
...
+		commonNameDetails := commonName.(map[string]interface{})
+		certificateCommonName := commonNameDetails["certificate_common_name"].(string)
+		certificateIssuerThumbprint := commonNameDetails["certificate_issuer_thumbprint"].(string)
+
+		commonName := servicefabric.ServerCertificateCommonName{
+			CertificateCommonName:       &certificateCommonName,
+			CertificateIssuerThumbprint: &certificateIssuerThumbprint,
+		}
+
+		commonNames = append(commonNames, commonName)
```
[the remark this answers is not in the document]
Comment (steve-hawkins) [reply azurerm/resource_arm_service_fabric_cluster.go]: removed length

File: azurerm/resource_arm_service_fabric_cluster.go [hunk 3]
```diff
@@ -644,6 +687,68 @@ func flattenServiceFabricClusterCertificate(input *servicefabric.CertificateDesc
...
+		for _, i := range *commonNames {
+			commonName := make(map[string]interface{})
+
+			commonName["certificate_common_name"] = *i.CertificateCommonName
+
+			if i.CertificateIssuerThumbprint != nil {
+				commonName["certificate_issuer_thumbprint"] = *i.CertificateIssuerThumbprint
+			}
+
+			common_names = append(common_names, commonName)
```
[the remark this answers is not in the document]
Comment (steve-hawkins) [reply azurerm/resource_arm_service_fabric_cluster.go]: removed length

**Comment by steve-hawkins** [comment]
@mbfrahry thanks for the feedback

out of interest, how come the failed test wasn't picked up in the CI build?

**Comment by mbfrahry** [comment]
We've got a couple of different CI systems in place for providers. Travis lints, vets, runs unit tests, etc and then we have TeamCity which runs our acceptance tests. We've got TeamCity hooked up privately for now so we can tailor what gets run on specific branches and so it doesn't get abused. There are plans to get the results more public facing but no details as of yet to share out.

**Comment by steve-hawkins** [comment]
@mbfrahry thanks for the info

let me know if there is anything else you wanted me to change as part of this PR

File: azurerm/resource_arm_service_fabric_cluster_test.go [hunk 4]
```diff
@@ -426,7 +506,54 @@ func TestAccAzureRMServiceFabricCluster_azureActiveDirectory(t *testing.T) {
...
+					testCheckAzureRMServiceFabricClusterExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "certificate.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "certificate.0.thumbprint", "33:41:DB:6C:F2:AF:72:C6:11:DF:3B:E3:72:1A:65:3A:F1:D4:3E:CD:50:F5:84:F8:28:79:3D:BE:91:03:C3:EE"),
+					resource.TestCheckResourceAttr(resourceName, "certificate.0.x509_store_name", "My"),
+					resource.TestCheckResourceAttr(resourceName, "azure_active_directory.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "fabric_settings.0.name", "Security"),
+					resource.TestCheckResourceAttr(resourceName, "fabric_settings.0.parameters.ClusterProtectionLevel", "EncryptAndSign"),
+					resource.TestCheckResourceAttr(resourceName, "management_endpoint", "https://example:19080"),
 				),
 			},
```
Comment (mbfrahry) [inline azurerm/resource_arm_service_fabric_cluster_test.go]: Let's add an import test step here to confirm that works correctly

File: azurerm/resource_arm_service_fabric_cluster_test.go [hunk 5]
```diff
@@ -751,6 +916,41 @@ resource "azurerm_service_fabric_cluster" "test" {
...
+}
+
+resource "azurerm_service_fabric_cluster" "test" {
+  name                = "acctest-%d"
+  resource_group_name = "${azurerm_resource_group.test.name}"
+  location            = "${azurerm_resource_group.test.location}"
+  reliability_level   = "Bronze"
+  upgrade_mode        = "Automatic"
+  vm_image            = "Windows"
+	management_endpoint = "http://example:80"
```
Comment (mbfrahry) [inline azurerm/resource_arm_service_fabric_cluster_test.go]: Let's fix the formatting here.

File: azurerm/resource_arm_service_fabric_cluster_test.go [hunk 6]
```diff
@@ -1059,11 +1349,53 @@ resource "azurerm_service_fabric_cluster" "test" {
...
+data "azurerm_client_config" "current" {}
+
+resource "azurerm_service_fabric_cluster" "test" {
+  name                = "acctest-%d"
+  resource_group_name = "${azurerm_resource_group.test.name}"
+  location            = "${azurerm_resource_group.test.location}"
+  reliability_level   = "Bronze"
+  upgrade_mode        = "Automatic"
+  vm_image            = "Windows"
+	management_endpoint = "https://example:19080"
```
Comment (mbfrahry) [inline azurerm/resource_arm_service_fabric_cluster_test.go]: And here

File: azurerm/resource_arm_service_fabric_cluster_test.go [hunk 7]
```diff
@@ -1059,11 +1349,53 @@ resource "azurerm_service_fabric_cluster" "test" {
...
+	
+  certificate {
+    thumbprint      = "33:41:DB:6C:F2:AF:72:C6:11:DF:3B:E3:72:1A:65:3A:F1:D4:3E:CD:50:F5:84:F8:28:79:3D:BE:91:03:C3:EE"
+    x509_store_name = "My"
+	}
```
Comment (mbfrahry) [inline azurerm/resource_arm_service_fabric_cluster_test.go]: And here

File: azurerm/resource_arm_service_fabric_cluster_test.go [hunk 8]
```diff
@@ -1059,11 +1349,53 @@ resource "azurerm_service_fabric_cluster" "test" {
...
+	
+  fabric_settings {
+		name = "Security"
```
Comment (mbfrahry) [inline azurerm/resource_arm_service_fabric_cluster_test.go]: And here

File: azurerm/resource_arm_service_fabric_cluster_test.go [hunk 9]
```diff
@@ -1059,11 +1349,53 @@ resource "azurerm_service_fabric_cluster" "test" {
...
+		
+    parameters = {
+      "ClusterProtectionLevel" = "EncryptAndSign"
+    }
+	}
```
Comment (mbfrahry) [inline azurerm/resource_arm_service_fabric_cluster_test.go]: And here

**Review by mbfrahry — CHANGES_REQUESTED** [review]
So close! Just some very minor test formatting fixes and a final test check and we're there.

**Comment by steve-hawkins** [comment]
@mbfrahry thanks, vscode auto-format gets me every time on the Terraform with golang files

File: azurerm/resource_arm_service_fabric_cluster_test.go (hunk 4 above)
Comment (steve-hawkins) [reply azurerm/resource_arm_service_fabric_cluster_test.go]: @mbfrahry apologies relatively new to this, do you mean this is different from the ImportState and ImportStateVerify?

Could you point me at an example I could follow?

Comment (mbfrahry) [reply azurerm/resource_arm_service_fabric_cluster_test.go]: My apologies @steve-hawkins. It's in there I just missed it.

**Review by mbfrahry — APPROVED** [review]

**Comment by mbfrahry** [comment]
No worries @steve-hawkins. @katbyte wrote a pretty neat tool that formats hcl. It's still a WIP but it should help out here going forward. https://github.com/katbyte/terrafmt. Other than that, this looks good! Thanks so much for your time here.

**Pull request merged**